lib/Runtime/Library/JavascriptGenerator.cpp

```diff
@@ -247,6 +262,419 @@ namespace Js
+            RecyclerNew(scriptContext->GetRecycler(), JavascriptExceptionObject, input, scriptContext, nullptr), _u("AsyncGenerator.prototype.throw"));
+    }
+
+
+    Var JavascriptGenerator::EntryAsyncGeneratorAwaitReject(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ScriptContext* scriptContext = function->GetScriptContext();
+        PROBE_STACK(scriptContext, Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+        AssertOrFailFastMsg(args.Info.Count > 1, "Should never call EntryAsyncGeneratorAwait without a parameter");
+
+        AsyncGeneratorNextProcessor* resumeNextReturnProcessor = VarTo<AsyncGeneratorNextProcessor>(function);
+
+        ResumeYieldData yieldData(args[1], RecyclerNew(scriptContext->GetRecycler(), JavascriptExceptionObject, args[1], scriptContext, nullptr));
+        resumeNextReturnProcessor->GetGenerator()->CallAsyncGenerator(&yieldData);
+        return scriptContext->GetLibrary()->GetUndefined();
+    }
+
+    Var JavascriptGenerator::EntryAsyncGeneratorAwaitRevolve(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ScriptContext* scriptContext = function->GetScriptContext();
+        PROBE_STACK(scriptContext, Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+        AssertOrFailFastMsg(args.Info.Count > 1, "Should never call EntryAsyncGeneratorAwait without a parameter");
+
+        AsyncGeneratorNextProcessor* resumeNextReturnProcessor = VarTo<AsyncGeneratorNextProcessor>(function);
+
+        ResumeYieldData yieldData(args.Values[1], nullptr);
+        resumeNextReturnProcessor->GetGenerator()->CallAsyncGenerator(&yieldData);
+        return scriptContext->GetLibrary()->GetUndefined();
+    }
+
+    void JavascriptGenerator::CallAsyncGenerator(ResumeYieldData* yieldData)
+    {
+        ScriptContext* scriptContext = this->GetScriptContext();
+        Var result = nullptr;
+        JavascriptExceptionObject *exception = nullptr;
+        yieldData->generator = this;
+        AssertMsg(isAsync, "Should not call CallAsyncGenerator on a non-async generator");
+        {
+            // RAII helper to set the state of the generator to completed if an exception is thrown
+            // or if the save state InterpreterStackFrame is never created implying the generator
+            // is JITed and returned without ever yielding.
+            class GeneratorStateHelper
+            {
+                JavascriptGenerator* g;
+                bool didThrow;
+            public:
+                GeneratorStateHelper(JavascriptGenerator* g) : g(g), didThrow(true) { g->SetState(GeneratorState::Executing); }
+                ~GeneratorStateHelper() { if (didThrow || g->frame == nullptr) {g->SetState(GeneratorState::Completed); } }
+                void DidNotThrow() { didThrow = false; }
```

Comment:
Hm. I'm not sure we can ever execute a generator without having created an interpreter frame instance. One design issue in the JIT scenario that we haven't fully worked out is the preservation of state between a yield and a next. We may need to have an interpreter instance to support this even after the generator has been jitted. For now, I would assume that an interpreter frame is always present.

Reply:
In that case I'll see if I can re-work the logic here a bit to simplify it - likely doing some unnecessary work.